### PR TITLE
fix(otel): extract Anthropic cache tokens from provider metadata for non-AI instrumentations

### LINKED ISSUE #13988: bug: Anthropic cache-read tokens dropped before Langfuse persists usage (Google ADK + LiteLLM)
### Describe the bug

Hey! Pretty sure I ran into the Python / Google-ADK cousin of langfuse/langfuse#13811 (same kind of thing, but that one's about the JS langchain integration).

My setup: a Google ADK agent running Claude (Sonnet) through LiteLLM, with Anthropic prompt caching turned on for the system prompt. For tracing I just create a `Langfuse()` client at startup so it registers the global OpenTelemetry tracer provider, and ADK's spans get picked up from there — I'm not making any manual generation calls on the main path.

The cache is clearly working: LiteLLM's response comes back with `cache_read_input_tokens` filled in on the cached turns. But by the time the generation lands in Langfuse, the usage only has `input` / `output` / `total` — no `cache_read` / `cache_creation` breakdown at all. So every cached token gets billed as fresh input, and since Anthropic only charges \~10% for cache reads, the cost Langfuse reports on cache-heavy turns ends up way higher than what I'm actually paying (several times over on turns where most of the prompt is cached).

For what it's worth, the in-process `langfuse.openai` wrapper handles this fine — its `_parse_usage` keeps `prompt_tokens_details` / `input_token_details`. So this looks specific to the OTel / ADK ingestion path, not the SDK in general.

**Expected:** the stored generation keeps the cache split (something like `input_cache_read`, ideally `input_cache_write` too) and Langfuse applies the discounted cache-read price.

**Actual:** the whole prompt is priced as uncached input, so cost is overcounted.

### Steps to reproduce

Rough repro — fire the same large system prompt twice so the second turn hits Anthropic's prompt cache:

```python
from google.adk.models.lite_llm import LiteLlm
from google.adk.plugins import BasePlugin

model = LiteLlm(
    model="anthropic/claude-sonnet-4-6",
    api_base="<your anthropic-compatible gateway>",
    api_key="<key>",
    drop_params=True,
    # litellm marks the system message ephemeral -> Anthropic prompt caching
    cache_control_injection_points=[
        {"location": "message", "role": "system", "control": {"type": "ephemeral"}}
    ],
)

# tiny plugin just to dump what ADK hands back
class UsageProbe(BasePlugin):
    async def after_model_callback(self, *, callback_context, llm_response):
        print("ADK usage_metadata:", llm_response.usage_metadata)
```

Wire that model into an agent, run the same prompt twice, then open the two generations in Langfuse.

Following the token count through the layers (same idea as langfuse/langfuse#13811):

* **LiteLLM raw response** → `cache_read_input_tokens` is there (a few thousand on turn 2). ✅
* **Langfuse stored usage** → only `input` / `output` / `total`, cache numbers gone. ❌

The one bit I haven't fully nailed down is the middle layer — whether ADK's `usage_metadata.cached_content_token_count` already comes back empty (→ google-adk isn't putting it on the span) or whether it's there and Langfuse's OTel ingestion just isn't mapping it (→ the direct Python-side equivalent of langfuse/langfuse#13811). A pointer on where to look would be great — happy to dig in once I know which side owns it.

### Langfuse Cloud or self-hosted?

Self-hosted

### If self-hosted, what version are you running?

*No response*

### SDK and integration versions

* `langfuse` (Python SDK): **3.8.1**
* `google-adk`: **1.31.1**
* `litellm`: **1.79.3**
* `opentelemetry-sdk`: **1.38.0**
* Model: `claude-sonnet-4-6` via the LiteLLM `anthropic/` provider, prompt caching enabled with `cache_control_injection_points` on the system message
* Python: **3.10**

### Additional information

Possibly related, possibly its own thing: on this ADK version I also see two token-bearing spans per LLM call (`call_llm` and `generate_content`), and both seem to get persisted as generations, which double-counts tokens independently of the cache issue. Happy to split that into a separate report if it's unrelated — just flagging it in case they share a code path.

### Are you interested in contributing a fix for this bug?

Yes

### PR DESCRIPTION
**Pull request by ZXT-zjbiliy**
## Description

When using Google ADK + LiteLLM as the LLM gateway, Anthropic cache tokens arrive via `ai.response.providerMetadata` but the instrumentation scope is `openinference.instrumentation.google_adk` instead of `ai`. The existing code only extracted Anthropic cache tokens when the scope was `ai`, causing cache-read and cache-creation tokens to be dropped.

Add `extractAnthropicCacheFromProviderMetadata()` to the generic usage extraction path so that Anthropic cache tokens are preserved regardless of the instrumentation scope.

Fixes #13988

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (tooling, dependencies, CI, workflows, repo upkeep, or other maintenance work)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (restructures existing code without changing behavior, e.g. simplify logic, split modules, reduce duplication)
- [ ] This change requires a documentation update

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- [x] I have read the [contributing guide](https://github.com/langfuse/langfuse/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my PR needs changes to the documentation
- [x] I have checked if my changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked if new and existing unit tests pass locally with my changes

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes a bug where Anthropic cache tokens (cache-read and cache-creation) were silently dropped when spans arrived from non-`ai`-scoped instrumentations like Google ADK + LiteLLM. The fix moves cache-token extraction into the generic usage path so it applies regardless of instrumentation scope.

- Adds `extractAnthropicCacheFromProviderMetadata()` that reads `ai.response.providerMetadata`, extracts Anthropic cache fields, and merges them into `usageDetails`. Freshly-set-field flags (`addedCacheRead`, `addedCacheCreation`, `addedCacheCreation5m`, `addedCacheCreation1h`) prevent double-subtracting from `input` when the generic `gen_ai.usage.*` path had already populated those fields.
- Adds two new test cases: one for the primary Google ADK + LiteLLM scenario, and one specifically covering the double-subtraction regression (both `gen_ai.usage.cache_read.input_tokens` and `ai.response.providerMetadata` carry the same cache-read value).
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the primary fix (cache tokens missing for Google ADK + LiteLLM spans) is correct and well-tested, with no regressions on the existing AI-SDK path.

The flag-based approach correctly prevents double-subtracting `input_cached_tokens` and `input_cache_creation` when the generic gen_ai.usage path already set them. Both new test cases pass and directly exercise the targeted regression. The one edge case noted (flat cache_creation pre-set by gen_ai.usage + duration breakdown from providerMetadata on the same span) is an unusual attribute combination unlikely to arise in the LiteLLM + Google ADK stack this PR targets.

No files require special attention for the primary fix; the logic around duration-specific cache breakdown in `OtelIngestionProcessor.ts` is worth a follow-up test if LiteLLM begins emitting both flat `gen_ai.usage.cache_creation.input_tokens` and duration-specific providerMetadata simultaneously.
</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[extractUsageDetails] --> B{instrumentationScope == 'ai'?}
    B -- Yes --> C[AI-SDK-specific path\n- reads providerMetadata\n- sets all cache fields\n- adjusts input]
    B -- No --> D[extractGenericGenAiUsageDetails]
    D --> E[Parse gen_ai.usage.* attrs\nSet input_cached_tokens\nSet input_cache_creation\nAdjust input = inputTokens - cacheRead - cacheCreation]
    E --> F[extractAnthropicCacheFromProviderMetadata]
    F --> G{providerMetadata present?}
    G -- No --> H[return]
    G -- Yes --> I[Parse anthropic.usage\nfrom providerMetadata]
    I --> J{input_cached_tokens\nalready set?}
    J -- No --> K[Set input_cached_tokens\naddedCacheRead=true]
    J -- Yes --> L[Skip - avoid double-subtract]
    K --> M{input_cache_creation\nalready set?}
    L --> M
    M -- No --> N[Set input_cache_creation\naddedCacheCreation=true]
    M -- Yes --> O[Skip]
    N --> P[Handle 5m/1h duration breakdown\nReduce input_cache_creation by 5m+1h]
    O --> P
    P --> Q[Re-derive input:\nSubtract only freshly-added fields]
    Q --> H
```
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 1 code review issue. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 1
packages/shared/src/server/otel/OtelIngestionProcessor.ts:2620-2660
**Duration-breakdown subtraction skipped when `input_cache_creation` was pre-populated**

When `gen_ai.usage.cache_creation.input_tokens` is already present on a span (so `addedCacheCreation = false`), and `ai.response.providerMetadata` additionally carries the duration-specific breakdown (`cache_creation.ephemeral_5m_input_tokens`), the function sets `addedCacheCreation5m = true` and later subtracts `input_cache_creation_5m` from `input`. But `input` was already reduced by the full `input_cache_creation` total (which includes those 5m tokens), so the 5m/1h amounts get subtracted a second time. Furthermore, `input_cache_creation` is never reduced to the remainder (the `if (addedCacheCreation)` guard prevents it), leaving it inflated instead of reflecting only the unclassified portion.

This only triggers when an instrumentation emits both a flat `gen_ai.usage.cache_creation.input_tokens` attribute *and* a providerMetadata payload with the duration breakdown — an unusual but plausible LiteLLM combination. Consider guarding the 5m/1h subtraction from `input` with an additional check that `addedCacheCreation` is also true (or that the total `input_cache_creation` was not already accounted for).

`````

</details>

<sub>Reviews (2): Last reviewed commit: ["test: add dual-source cache token test t..."](https://github.com/langfuse/langfuse/commit/844a3da39070bd0e15939cd1d7c3e12f80de4118) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36071066)</sub>

### CODE REVIEW

File: packages/shared/src/server/otel/OtelIngestionProcessor.ts [hunk 1]
```diff
@@ -2537,9 +2537,121 @@ export class OtelIngestionProcessor {
...
+      ) {
+        usageDetails["input"] = Math.max(
+          usageDetails["input"] -
+            (usageDetails["input_cached_tokens"] ?? 0) -
+            (usageDetails["input_cache_creation"] ?? 0) -
+            (usageDetails["input_cache_creation_5m"] ?? 0) -
+            (usageDetails["input_cache_creation_1h"] ?? 0),
+          0,
+        );
+      }
```
[the remark this answers is not in the document]
Comment (ZXT-zjbiliy) [reply packages/shared/src/server/otel/OtelIngestionProcessor.ts]: I've added `addedCacheRead` / `addedCacheCreation` / `addedCacheCreation5m` / `addedCacheCreation1h` flags that track which fields this function actually set. The re-derivation block now only subtracts fields that were freshly populated, so when `extractGenericGenAiUsageDetails` already computed `input = inputTokens − cacheReadTokens − cacheCreationTokens`, those amounts won't be subtracted again.

File: web/src/__tests__/server/api/otel/otelMapping.servertest.ts [hunk 2]
```diff
@@ -7203,6 +7203,117 @@ describe("OTel Resource Span Mapping", () => {
...
+        500,
+      );
+
+      // input = 5000 - 1500 (cached) - 500 (creation remainder) - 1000 (5m) - 500 (1h) = 1500
+      expect(observationEvent?.body.usageDetails.input).toBe(1500);
+      expect(observationEvent?.body.usageDetails.output).toBe(200);
+    });
+
     it("should extract all Bedrock cache token types from Vercel AI SDK provider metadata", async () => {
       const traceId = "abcdef1234567890abcdef1234567890";
```
[the remark this answers is not in the document]
Comment (ZXT-zjbiliy) [reply web/src/__tests__/server/api/otel/otelMapping.servertest.ts:7431]: Added a test case `"should not double-subtract when both gen_ai.usage and providerMetadata carry cache tokens"` that sends `gen_ai.usage.cache_read.input_tokens=3000` and `ai.response.providerMetadata` with `cache_read_input_tokens=3000` + `cache_creation_input_tokens=1000` on the same span. It asserts `input = 8000 − 3000 − 1000 = 4000` (not 0), directly verifying the regression.

**Comment by nimarb** [comment]
@claude review
